### DISCUSSION TITLE
case and case/all returning values

### DISCUSSION THREAD

**Discussion by gor77** [red/red#5519] [Q&A] [answered]
This is related to issue https://github.com/red/red/issues/5517 , case/all returns _false_ instead of _none_. 
This bug  made me believe that evaluation of **conditions itself** will sort of speak "stay in that block". 

Following that path it's so easy to create wrong mental model about how things work. 
Docs are not clear at all about what actually happens with results of **condition evaluation**.

It might seem trivial to more experienced users, but given the fact Red is a functional language
full of expressions, and also being very much different from practically
everything out there except Rebol :), I think all this makes learning Red much harder than it should be. 

So, just to make things clear and try to put some clear definitions
that will hopefully help newcomers to better understand *case* and *case/all*:

1. *Case* will execute just the first block where corresponding condition is _true_.
   It will stop there, and no further evaluations will be done. If there are no conditions that evaluate to _true_, 
   *case* will return _none_.

2. *Case/all* will execute **each block** where corresponding condition is _true_. 
   If there are no conditions that evaluate to _true_, _case/all_ will return _none_. 

In other words, it's safe to think about _case_ (but not _case all_) as a bunch of if - else if... statements from other languages.
And _case/all_ is basically just a bunch of _if_ statements. (But _Red's if statements_ - as there is no else block in Red's if statement). 

So - is this all correct or am I missing something here?

**Comment by greggirwin** [comment, marked as answer]
We run up against this issue  a lot, because doc strings have to be short to be practical, but we don't have a full ecosystem to link to, with info like https://www.red-by-example.org/#case provides. With some things, no matter how we describe them, they may be confusing. that's where examples really help. Look at `switch` for a close relative of `case`, and how we explain their similarities and differences.

If you can formulate new doc strings for those two funcs that people agree are better than what we have now, which will take some chat time, then you can PR it.

> In other words, it's safe to think about case (but not case all) as a bunch of if - else if... statements from other languages.
And case/all is basically just a bunch of if statements. (But Red's if statements - as there is no else block in Red's if statement).

Yes, that's a good analogy. As an historical side note, Rebol's `if` had an `else` refinement, but it was removed because `either` is the better solution there.

**Reply by gor77** [reply]
Then maybe it would make sense to add just a single line to the description, something like this:

for _case_:
Returns the result of evaluated block, or none if no conditions have been satisfied.

for _case/all_:
Returns the result of last evaluated block, or none if no conditions have been satisfied.

I guess that should be enough (hopefully) to make things clear, as the return value 
can be only one of two things: the result of last evaluated block, or _none_.

**Reply by greggirwin** [reply]
OK, now take a crack at formulating the _actual_ doc strings for `case` and `switch` and look at `if` for consistency as well. I'll do the same here. It's a good exercise. See if you can keep the doc string under ~70 chars, which is sort of an informal limit we keep in mind.

**Reply by gor77** [reply]
Well now when I know all the constraints, yes...it is hard if not impossible. 
70 chars is extremely limiting.

Maybe something meaningful could be done in around 130-140 chars. Per entry, probably...

`Case`:
{Evaluates block after first true condition. 
Returns value of the evaluated block, or `none `if no conditions have been satisfied.}


`Case/all`:
{Evaluates each block where condition is true. 
Returns value of the last evaluated block, or `none `if no conditions have been satisfied.}

No idea how to do it shorter except maybe to place common behavior in a single place:
"Both case and case/all return none if no conditions have been satisfied", so there is 
no need to repeat it. 

Anyways, my opinion is that the bare minimum for every description should be 

1. what does this thing evaluate to (what's the return value)
2. does it modify its argument

In my mind, these things are crucial.

This is now going in slightly different direction, but if we can agree these things things should be in the description, 
then maybe it would make sense to adopt some kind of convention symbols which won't take much space, but would 
make clear that operation is destructive, for things like `insert`, `replace`, `take `etc. 

Or maybe (I don't know if I'm going too far with this, because I have no idea how difficult would it be to pull this off), 
to create another datatype, something like destructive!, or maybe just D!
and then you waste only 2 characters for conveying very important information. 

On top of that, this important info would be contained in the language itself. This way it would be very easy to 
check programmatically if operation is destructive, and future types and their documentation would also be safe from
somebody forgetting to write that in the description. 

For example, right now, without testing I can't really be sure whether `insert `is destructive or not. 
And I always forget these things.

**Reply by greggirwin** [reply]
It's tricky all right.

If we include `Returns value of the evaluated block`, should we also add that to `[if either switch for* ...]`? The idea that the last expression evaluated is returned is fundamental. For `if/unless` it's nice that they say `else returns NONE`, but is harder to formulate well for more complex funcs.

The immutable variant option comes up from time to time, but seems much worse than the current approach, which is:
- Assume series funcs modify them, and use `copy` when you don't want that.
- For things where that may not be clear, or other args are modified, include `(modified)` in the doc string for those args.
The reason you can't just take ideas from other langs is that a language is a whole, when considered that way. Suddenly you have line noise for every instance of the most common functions for their most common use case. The concept that Redbol langs use series references by default, rather than copying them, is a central point of design and understanding. 

Other issues are that if it's just a convention, you can't trust it from a formality standpoint. If you make it a new datatype, that's a lot of work, will very likely cause a whole lot of pain when you have to figure out compatibility or how values are coerced when they interact. e.g. what happens if you append a mutable series to an immutable one, or vice versa? Also `D!` is pretty meaningless. :^)

When people are new to a Redbol lang (Red, Rebol, or a more fringe variant), they often think things need to be changed. And some things can be improved. But over time, as with many langs, use becomes more natural and the design makes sense. Consider that Rebol was designed over a 20-year period by a great designer/developer. Now it's had another 25 years of scrutiny by other amazing designers and developers. Doesn't meant it's perfect, but it is what it is. We can change things and make it different, but different doesn't mean better.

**Reply by hiiamboris** [reply]
Docstrings are a hint, not a reference documentation. So this issue will be solved once we get to that stage.
